agent/csi/volumes.go

```diff
@@ -0,0 +1,248 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu        sync.RWMutex                     // To sync map "m" and "pluginMap"
+	tryMu     sync.RWMutex                     // To sync between tryAddVolume() and tryRemoveVolume()
+	m         map[string]*api.VolumeAssignment // Map between VolumeID and VolumeAssignment
+	pluginMap map[string]*NodePlugin           // Map between Driver Name and NodePlugin
+}
+
+const maxRetries int = 20
+
+const initialBackoff = 1 * time.Millisecond
+
+// NewManager returns a place to store volumes.
+func NewManager() exec.VolumesManager {
+	return &volumes{
+		m:         make(map[string]*api.VolumeAssignment),
+		pluginMap: make(map[string]*NodePlugin),
+	}
+}
+
+// Get returns a volume published path for the provided volume ID.  If the volume doesn't exist, returns empty string.
+func (r *volumes) Get(volumeID string) (string, error) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	if plugin, ok := r.pluginMap[volumeID]; ok {
```

Comment:
This won't work, because the pluginMap maps the driver name to the plugin, not the volumeID to the plugin. It's probably better to keep track of published paths in the volume manager, instead of in the plugin, so that you can directly look them up without having to resolve which plugin a given volume uses.